archive/100-theorems-list/16_abel_ruffini.lean

```diff
@@ -0,0 +1,206 @@
+    rw this,
+    exact lt_of_le_of_lt degree_C_le (with_bot.coe_lt_coe.mpr (nat.zero_lt_bit1 2)) },
+  rw [degree_sub_eq_left_of_degree_lt, degree_X_pow],
+  rw [degree_X_pow, ←pow_one X],
+  exact lt_of_le_of_lt (degree_C_mul_X_pow_le _ _)
+    (with_bot.coe_lt_coe.mpr (nat.one_lt_bit1 two_ne_zero)),
+end
+
+lemma nat_degree_Phi : (Φ R a b).nat_degree = 5 :=
+nat_degree_eq_of_degree_eq_some (degree_Phi a b)
+
+lemma leading_coeff_Phi : (Φ R a b).leading_coeff = 1 :=
+by rw [leading_coeff, nat_degree_Phi, coeff_five_Phi]
+
+lemma monic_Phi : (Φ R a b).monic :=
+leading_coeff_Phi a b
+
+lemma irreducible_Phi (p : ℕ) (hp : p.prime) (hpa : p ∣ a) (hpb : p ∣ b) (hp2b : ¬ (p ^ 2 ∣ b)) :
+  irreducible (Φ ℚ a b) :=
+begin
+  rw [←map_Phi a b (int.cast_ring_hom ℚ), ←is_primitive.int.irreducible_iff_irreducible_map_cast],
+  apply irreducible_of_eisenstein_criterion,
+  { rwa [ideal.span_singleton_prime (int.coe_nat_ne_zero.mpr hp.ne_zero),
+      int.prime_iff_nat_abs_prime] },
+  { rw [leading_coeff_Phi, ideal.mem_span_singleton],
+    exact_mod_cast mt nat.dvd_one.mp (hp.ne_one) },
+  { intros n hn,
+    rw ideal.mem_span_singleton,
+    rw [degree_Phi, with_bot.coe_lt_coe] at hn,
+    interval_cases n with hn;
+    simp [Φ, coeff_X_pow, coeff_C, int.coe_nat_dvd.mpr, hpb, hpa, -ring_hom.eq_int_cast] },
+  { simp only [degree_Phi, ←with_bot.coe_zero, with_bot.coe_lt_coe, nat.succ_pos'] },
+  { rwa [coeff_zero_Phi, pow_two, ideal.span_singleton_mul_span_singleton, ←pow_two,
+      ideal.mem_span_singleton, ←int.coe_nat_pow],
+    norm_num,
+    exact mt int.coe_nat_dvd.mp hp2b },
+  all_goals { exact polynomial.monic.is_primitive (monic_Phi a b) },
+end
+
+lemma real_roots_Phi_le : fintype.card ((Φ ℚ a b).root_set ℝ) ≤ 3 :=
+begin
+  rw [←map_Phi a b (algebra_map ℤ ℚ), Φ, ←one_mul (X ^ 5), ←C_1],
+  refine (card_root_set_le_derivative _).trans
+    (nat.succ_le_succ ((card_root_set_le_derivative _).trans (nat.succ_le_succ _))),
+  suffices : ((C ((algebra_map ℤ ℚ) 20) * X ^ 3).root_set ℝ).subsingleton,
+  { norm_num [fintype.card_le_one_iff_subsingleton, ← mul_assoc, *] at * },
+  rw root_set_C_mul_X_pow; norm_num,
+end
+
+lemma real_roots_Phi_ge_aux' (hab : b < a) : a ^ 2 + b ≤ a ^ 5 :=
+begin
+  refine le_trans ((nat.add_le_add_iff_le_right 1 _ _).mp _) (nat.pow_le_pow_of_le_right
+    (nat.one_le_of_lt hab) (bit1_le_bit1.mpr one_le_two)),
+  rw [add_assoc],
+  apply (add_le_add (le_refl (a ^ 2)) (nat.succ_le_iff.mpr hab)).trans,
+  suffices : ∀ c : ℤ, 0 ≤ c → c ^ 2 + c ≤ c ^ 3 + 1,
+  { specialize this a (int.coe_nat_nonneg a),
+    norm_cast at this,
+    exact this },
+  intros c hc,
+  rw [←sub_nonneg, show c ^ 3 + 1 - (c ^ 2 + c) = (c - 1) ^ 2 * (c + 1), by ring],
+  exact mul_nonneg (pow_two_nonneg _) (add_nonneg hc zero_le_one),
+end
+
+lemma real_roots_Phi_ge_aux (hab : b < a) (f : ℝ → ℝ) (f_cont : continuous f)
+  (f_def : ∀ x : ℝ, f x = x ^ 5 - a * x + b) : ∃ x y : ℝ, x ≠ y ∧ f x = 0 ∧ f y = 0 :=
```

Comment:
At first glance this looks strange to me. Why wouldn't you state this as
```
lemma real_roots_Phi_ge_aux (hab : b < a) : 
  ∃ x y : ℝ, x ≠ y ∧ x ^ 5 - a * x + b = 0 ∧ y ^ 5 - a * y + b = 0 :=
```
The fact that 
`λ x, x ^ 5 - a * x + b` is `continuous` is trivial:
```
have : continuous (λ x, x ^ 5 - a * x + b) := by continuity,
```

Reply:
(And if you want to define `f` for convenience you can use a `let` statement)

Reply:
Good point. I'll take a look.

Reply:
I've changed the lemma to
```
lemma real_roots_Phi_ge_aux (hab : b < a) :
  ∃ x y : ℝ, x ≠ y ∧ aeval x (Φ ℚ a b) = 0 ∧ aeval y (Φ ℚ a b) = 0 :=
```
since that's actually what's needed.